What is an API
“An application programming interface (API) is a computing interface that defines interactions between multiple software or mixed hardware-software intermediaries.” – Wikipedia.org

In [2]:
import pandas_datareader as pdr
import pandas as pd
import datetime as dt

In [3]:
start  = dt.datetime(2010,1,1)
tcs = pdr.get_data_yahoo("TCS.NS", start)

In [4]:
tcs.head()

High         Low        Open       Close     Volume  \
Date                                                                    
2010-01-04  379.450012  374.524994  377.399994  375.825012  1963682.0   
2010-01-05  379.774994  373.274994  377.500000  375.924988  2014488.0   
2010-01-06  376.100006  366.500000  375.924988  367.424988  3349176.0   
2010-01-07  369.700012  355.500000  367.750000  357.200012  6474892.0   
2010-01-08  359.250000  348.250000  358.000000  349.899994  6048178.0   

             Adj Close  
Date                    
2010-01-04  299.222321  
2010-01-05  299.302002  
2010-01-06  292.534454  
2010-01-07  284.393707  
2010-01-08  278.581635

In [5]:
tcs.tail()

High          Low     Open        Close     Volume  \
Date                                                                    
2022-03-08  3609.000000  3460.000000  3460.00  3599.949951  3300992.0   
2022-03-09  3654.000000  3597.050049  3605.00  3632.600098  2719654.0   
2022-03-10  3684.000000  3615.000000  3684.00  3620.949951  2005441.0   
2022-03-11  3639.850098  3592.850098  3603.25  3599.149902  1623423.0   
2022-03-11  3639.850098  3592.850098  3603.25  3599.149902  1622215.0   

              Adj Close  
Date                     
2022-03-08  3599.949951  
2022-03-09  3632.600098  
2022-03-10  3620.949951  
2022-03-11  3599.149902  
2022-03-11  3599.149902

In [7]:
tickers = ['TCS.NS','INFY.NS','RELIANCE.NS','VEDL.NS']
start = dt.datetime(2010,1,1)
data = pdr.get_data_yahoo(tickers,start)

In [8]:
data.head()

Attributes   Adj Close                                           Close  \
Symbols         TCS.NS     INFY.NS RELIANCE.NS     VEDL.NS      TCS.NS   
Date                                                                     
2010-01-04  299.222443  252.160767  484.266113  216.092972  375.825012   
2010-01-05  299.302063  253.005280  482.104889  217.685562  375.924988   
2010-01-06  292.534576  249.313522  489.894531  215.362030  367.424988   
2010-01-07  284.393677  243.710648  498.021881  212.933945  357.200012   
2010-01-08  278.581543  237.837646  496.716125  211.054199  349.899994   

Attributes                                            High              ...  \
Symbols        INFY.NS RELIANCE.NS     VEDL.NS      TCS.NS     INFY.NS  ...   
Date                                                                    ...   
2010-01-04  326.575012  532.700500  413.850006  379.450012  328.750000  ...   
2010-01-05  327.668762  530.323059  416.899994  379.774994  328.837494  ...   
2010-01-06  322.887512  538.891846  412.450012  376.100006  335.375000  ...   
2010-01-07  315.631256  547.832092  407.799988  369.700012  323.325012  ...   
2010-01-08  308.024994  546.395691  404.200012  359.250000  315.412506  ...   

Attributes         Low                    Open                          \
Symbols    RELIANCE.NS     VEDL.NS      TCS.NS     INFY.NS RELIANCE.NS   
Date                                                                     
2010-01-04  506.127411  396.000000  377.399994  326.250000  540.427307   
2010-01-05  527.697937  414.200012  377.500000  327.500000  569.551208   
2010-01-06  530.298279  410.000000  375.924988  335.375000  534.879883   
2010-01-07  533.938782  406.299988  367.750000  323.325012  538.891846   
2010-01-08  542.854309  401.000000  358.000000  314.875000  548.797974   

Attributes                 Volume                                     
Symbols        VEDL.NS     TCS.NS     INFY.NS RELIANCE.NS    VEDL.NS  
Date                                                                  
2010-01-04  411.100006  1963682.0   4069264.0  35372156.0  4639413.0  
2010-01-05  417.700012  2014488.0   6895528.0   9872785.0  5319410.0  
2010-01-06  418.149994  3349176.0   6817288.0  10933743.0  3891427.0  
2010-01-07  413.799988  6474892.0  10892600.0  12090336.0  4389842.0  
2010-01-08  408.700012  6048178.0  12649312.0   6973331.0  2936406.0  

[5 rows x 24 columns]

In [9]:
data.tail()

Attributes    Adj Close                                              Close  \
Symbols          TCS.NS      INFY.NS  RELIANCE.NS     VEDL.NS       TCS.NS   
Date                                                                         
2022-03-08  3599.949951  1778.199951  2235.500000  368.350006  3599.949951   
2022-03-09  3632.600098  1813.349976  2354.100098  371.250000  3632.600098   
2022-03-10  3620.949951  1825.900024  2392.250000  378.250000  3620.949951   
2022-03-11  3599.149902  1822.000000  2399.149902  378.200012  3599.149902   
2022-03-11  3599.149902  1822.000000  2399.149902  378.200012  3599.149902   

Attributes                                               High               \
Symbols         INFY.NS  RELIANCE.NS     VEDL.NS       TCS.NS      INFY.NS   
Date                                                                         
2022-03-08  1778.199951  2235.500000  381.350006  3609.000000  1800.000000   
2022-03-09  1813.349976  2354.100098  371.250000  3654.000000  1831.599976   
2022-03-10  1825.900024  2392.250000  378.250000  3684.000000  1844.750000   
2022-03-11  1822.000000  2399.149902  378.200012  3639.850098  1831.400024   
2022-03-11  1822.000000  2399.149902  378.200012  3639.850098  1831.400024   

Attributes  ...          Low                 Open                            \
Symbols     ...  RELIANCE.NS     VEDL.NS   TCS.NS      INFY.NS  RELIANCE.NS   
Date        ...                                                               
2022-03-08  ...  2180.000000  372.500000  3460.00  1718.449951  2211.500000   
2022-03-09  ...  2234.149902  365.450012  3605.00  1791.000000  2236.000000   
2022-03-10  ...  2360.000000  363.450012  3684.00  1844.750000  2400.000000   
2022-03-11  ...  2376.050049  376.200012  3603.25  1814.000000  2376.050049   
2022-03-11  ...  2376.050049  376.200012  3603.25  1814.000000  2376.050049   

Attributes                 Volume                                      
Symbols        VEDL.NS     TCS.NS     INFY.NS RELIANCE.NS     VEDL.NS  
Date                                                                   
2022-03-08  396.399994  3300992.0  10984500.0   9825043.0  19093725.0  
2022-03-09  369.950012  2719654.0  10193515.0   9421163.0  10587981.0  
2022-03-10  371.500000  2005441.0   6547182.0   8297995.0  12346925.0  
2022-03-11  377.799988  1623423.0   4758657.0   5656799.0   9045446.0  
2022-03-11  377.799988  1622215.0   4758444.0   5656056.0   9040767.0  

[5 rows x 24 columns]

Project – Normalize and Calculate Return of Portfolio

normalization of ratings means adjusting values measured on different scales to a notionally common scale, often prior to averaging.”

In [10]:
data = data['Adj Close']

In [11]:
data.head()

Symbols         TCS.NS     INFY.NS  RELIANCE.NS     VEDL.NS
Date                                                       
2010-01-04  299.222443  252.160767   484.266113  216.092972
2010-01-05  299.302063  253.005280   482.104889  217.685562
2010-01-06  292.534576  249.313522   489.894531  215.362030
2010-01-07  284.393677  243.710648   498.021881  212.933945
2010-01-08  278.581543  237.837646   496.716125  211.054199

Adj Close = "The adjusted closing price amends a stock's closing price to reflect that stock's value after accounting for any corporate actions. It is often used when examining historical returns or doing a detailed analysis of past performance."

In [12]:
norm = data/data.iloc[0]

In [23]:
norm.head(10)

Symbols       TCS.NS   INFY.NS  RELIANCE.NS   VEDL.NS
Date                                                 
2010-01-04  1.000000  1.000000     1.000000  1.000000
2010-01-05  1.000266  1.003349     0.995537  1.007370
2010-01-06  0.977649  0.988709     1.011623  0.996617
2010-01-07  0.950442  0.966489     1.028405  0.985381
2010-01-08  0.931018  0.943198     1.025709  0.976682
2010-01-11  0.950176  0.952940     1.005579  1.000241
2010-01-12  0.997871  0.990182     1.008229  0.977528
2010-01-13  1.049092  1.027138     1.012180  0.976562
2010-01-14  1.037517  1.029530     1.042956  1.012444
2010-01-15  1.052884  1.024190     1.032775  0.994080

As you see, all the start prices are adjusted to 1. This is the normalization

Then the relative change can be seen from day 2010-01-05 and forward

In [20]:
portfolio = [.25,.25,.25,.25]

In [21]:
weights = (norm*portfolio)

In [24]:
weights.head(10)

Symbols       TCS.NS   INFY.NS  RELIANCE.NS   VEDL.NS
Date                                                 
2010-01-04  0.250000  0.250000     0.250000  0.250000
2010-01-05  0.250067  0.250837     0.248884  0.251842
2010-01-06  0.244412  0.247177     0.252906  0.249154
2010-01-07  0.237611  0.241622     0.257101  0.246345
2010-01-08  0.232755  0.235800     0.256427  0.244171
2010-01-11  0.237544  0.238235     0.251395  0.250060
2010-01-12  0.249468  0.247546     0.252057  0.244382
2010-01-13  0.262273  0.256784     0.253045  0.244140
2010-01-14  0.259379  0.257383     0.260739  0.253111
2010-01-15  0.263221  0.256048     0.258194  0.248520

we get a weighted distribution of our portfolio. Starting with 25% for each ticker

In [25]:
weights['Total'] = (norm*portfolio).sum(axis=1)

In [26]:
weights.head(10)

Symbols       TCS.NS   INFY.NS  RELIANCE.NS   VEDL.NS     Total
Date                                                           
2010-01-04  0.250000  0.250000     0.250000  0.250000  1.000000
2010-01-05  0.250067  0.250837     0.248884  0.251842  1.001631
2010-01-06  0.244412  0.247177     0.252906  0.249154  0.993649
2010-01-07  0.237611  0.241622     0.257101  0.246345  0.982680
2010-01-08  0.232755  0.235800     0.256427  0.244171  0.969152
2010-01-11  0.237544  0.238235     0.251395  0.250060  0.977234
2010-01-12  0.249468  0.247546     0.252057  0.244382  0.993453
2010-01-13  0.262273  0.256784     0.253045  0.244140  1.016243
2010-01-14  0.259379  0.257383     0.260739  0.253111  1.030612
2010-01-15  0.263221  0.256048     0.258194  0.248520  1.025982

we have the sum of our weighted sum of how the portfolio evolves in the Total column

Let’s assume we invested 100,000 Rupees. This can be modeled quite easy as you see here.

In [29]:
(weights*100000).head(10)

Symbols           TCS.NS       INFY.NS   RELIANCE.NS       VEDL.NS  \
Date                                                                 
2010-01-04  25000.000000  25000.000000  25000.000000  25000.000000   
2010-01-05  25006.652272  25083.727630  24888.427855  25184.248280   
2010-01-06  24441.229562  24717.715378  25290.564310  24915.436656   
2010-01-07  23761.058350  24162.229008  25710.134751  24634.529171   
2010-01-08  23275.455254  23579.961476  25642.725759  24417.059641   
2010-01-11  23754.395879  23823.492606  25139.470302  25006.037339   
2010-01-12  24946.786925  24754.550634  25205.718183  24438.206214   
2010-01-13  26227.294437  25678.447050  25304.507087  24414.039206   
2010-01-14  25937.932085  25738.251204  26073.910453  25311.103005   
2010-01-15  26322.101422  25604.761410  25819.379949  24852.002232   

Symbols             Total  
Date                       
2010-01-04  100000.000000  
2010-01-05  100163.056037  
2010-01-06   99364.945907  
2010-01-07   98267.951280  
2010-01-08   96915.202129  
2010-01-11   97723.396126  
2010-01-12   99345.261956  
2010-01-13  101624.287780  
2010-01-14  103061.196747  
2010-01-15  102598.245014

Our portfolio looks like this in the end

In [31]:
(weights*100000).tail(10)

Symbols            TCS.NS        INFY.NS    RELIANCE.NS       VEDL.NS  \
Date                                                                    
2022-02-28  296952.989218  170089.899264  121810.611156  42497.423892   
2022-03-02  296280.408583  168820.874851  123823.966980  43279.651153   
2022-03-03  296134.196288  170610.400537  122778.570687  44246.267918   
2022-03-04  294459.037457  170853.308393  120055.377872  41726.367473   
2022-03-07  291162.978264  172494.119430  115615.670156  43290.829057   
2022-03-08  300775.396355  176296.254879  115406.588376  42614.760100   
2022-03-09  303503.312265  179781.137251  121529.260106  42950.263133   
2022-03-10  302529.943892  181025.387992  123498.730057  43760.099744   
2022-03-11  300708.552369  180638.727483  123854.934123  43754.316609   
2022-03-11  300708.552369  180638.727483  123854.934123  43754.316609   

Symbols             Total  
Date                       
2022-02-28  631350.923530  
2022-03-02  632204.901567  
2022-03-03  633769.435430  
2022-03-04  627094.091195  
2022-03-07  622563.596906  
2022-03-08  635092.999710  
2022-03-09  647763.972755  
2022-03-10  650814.161685  
2022-03-11  648956.530584  
2022-03-11  648956.530584

We have that 2022-03-11 we will have 648956.53 in total.

In [32]:
(weights['Total']*100000).iloc[-1]

648956.5305837922

Hence, this shows that if we invested 100,000 Rupees the opening day of 2010 in a portfolio consisting of 25% of each ticker, then we would have 648956.5305837922 Rupess on March 11th, 2022.